docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,679 @@
+until the pending operation is finished.
+
+**Assumptions:**
+
+- When enclave is loaded, the OS reserves the whole address range covered by ELRANGE.
+It is assumed the host app will not remap any part of this reserved range.
+- When an enclave is loaded with base address at zero, only partial ELRANGE may be
+  reserved by the OS. In that case, the EMM will assume the partial ELRANGE as valid reserved
+  range for use inside the enclave.
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low.
+The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave
+  and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions
+  to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure
+itself would not cause #GP, and only register a #PF handler with the enclave global exception
+handler registry through the runtime abstraction layer. Clients wish to handle #GP can register
+its own exception handler with the global handler registry.
+
+Public APIs
+-----------------
+
+### sgx_mm_alloc
+
+Allocate a new memory region inside enclave and optionally register a custom page fault handler
+for the region
+
+```
+/**
+ * Page fault (#PF) info reported in the SGX SSA MISC region.
+ */
+typedef struct _sgx_pfinfo
+{
+    uint64_t maddr; //address for #PF
+    union _pfec
+    {
+        uint32_t errcd;
+        struct
+        {                     //PFEC bits
+            uint32_t p : 1;   // P flag
+            uint32_t rw : 1;  // RW access flag, 0 for read, 1 for write
+            uint32_t : 13;    // U/S, I/O, PK and reserved bits not relevant for SGX PF
+            uint32_t sgx : 1; // SGX bit
+            uint32_t : 16;    // reserved bits
+        };
+    } pfec;
+    uint32_t reserved;
+} sgx_pfinfo;
+
+/**
+ * Custom page fault (#PF) handler, do usage specific processing upon #PF,
+ * e.g. loading data and verify its trustworthiness, then call sgx_mm_commit_data
+ * to explicitly EACCEPTCOPY data.
+ * This custom handler is passed into sgx_mm_alloc, and associated with the
+ * newly allocated region. The memory manager calls the handler when a #PF
+ * happens in the associated region. The handler may invoke abort() if it
+ * determines the exception is invalid based on certain internal states
+ * it maintains.
+ *
+ * @param[in] pfinfo, info reported in the SSA MISC region for page fault.
+ * @param[in] private_data, private data provided by handler in sgx_mm_alloc call.
+ * @retval EXCEPTION_CONTINUE_EXECUTION,  success on handling the exception.
+ * @retval EXCEPTION_CONTINUE_SEARCH, Exception not handled and should be passed
```

Comment:
```suggestion
 * @retval EXCEPTION_CONTINUE_EXECUTION Success on handling the exception.
 * @retval EXCEPTION_CONTINUE_SEARCH Exception not handled and should be passed.
```

nit: Don't put comma after value name in doxygen comments, and be consistent in terms of whether the next word is capitalized or not, and whether the phrase ends in punctuation